crates/proto/src/udp/udp_stream.rs

```diff
@@ -239,58 +244,58 @@ impl<P: RuntimeProvider> Future for NextRandomUdpSocket<P> {
     /// polls until there is an available next random UDP port,
     /// if no port has been specified in bind_addr.
     fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
-        if self.bind_address.port() == 0 {
-            // Per RFC 6056 Section 3.2:
-            //
-            // As mentioned in Section 2.1, the dynamic ports consist of the range
-            // 49152-65535.  However, ephemeral port selection algorithms should use
-            // the whole range 1024-65535.
-            let rand_port_range = Uniform::new_inclusive(1_024, u16::MAX);
-            let mut rand = rand::thread_rng();
-
-            for attempt in 0..10 {
-                let port = rand_port_range.sample(&mut rand);
-                let bind_addr = SocketAddr::new(self.bind_address.ip(), port);
-
-                // TODO: allow TTL to be adjusted...
-                // TODO: this immediate poll might be wrong in some cases...
-                match self
-                    .provider
-                    .bind_udp(bind_addr, self.name_server)
-                    .as_mut()
-                    .poll(cx)
-                {
+        let this = self.get_mut();
```

Comment:
Are you sure this is thread safe? I think it's better to just use `Option::take`. Also, below, after success, `attempted` is not reset.

Reply:
I'm not sure so I'm asking you. Thread safety means that the state is taken here, so even if other threads run this code at the same time, there will be no problem with this call.

Reply:
This takes a mutable reference as `self`, so no two threads can run this at the same time.